### PR TITLE
Bytearray iadd

### PR DESCRIPTION
**Pull request by devendor**
msgpack/fallback.py which newer pip depends on uses `bytearray += <type memorview>` which is supported by py2.x.

Note that it does not support += of the older <type buffer> most likely due to the fact that buffer can contain encoded Unicode whereas memoryview should throw a TypeError and force explicit encoding.  The typeerror is covered in a separate test/patch/pull-request.

### CODE REVIEW

**Comment by jeff5** [comment]
@devendor : You're right that += should work for a memoryview, but this isn't quite the way to do it. We should make `__iadd__` accept any object that satisfies the buffer protocol. We would then not need the clause for `PyString`, as the inner workings of setslice have it covered, and will even reject `PyUnicode` for us.

I wondered if this was the only place lacking in this way. I see that the comparison operators (actually implemented in `BaseBytes`) are compatible with the buffer protocol already. These offer a good model to copy. (Note the use of try-with-resources.)

I saw also that I have left a note in `__add__` to do the same there, but never followed up. Would you like to address that along with `__add__`?

File: src/org/python/core/PyByteArray.java [hunk 1]
```diff
@@ -1325,6 +1325,8 @@ final synchronized PyObject bytearray___iadd__(PyObject o) {
         } else if (oType == PyString.TYPE) {
             // Will fail if somehow not 8-bit clean
             setslice(size, size, 1, (PyString)o);
+        } else if (o instanceof PyMemoryView){
```
[the remark this answers is not in the document]
Comment (devendor) [reply src/org/python/core/PyByteArray.java]: @jeff5 msgpack is used in newer pip, so it's essential python functionality going forward.

[the remark this answers is not in the document]
Comment (jeff5) [reply src/org/python/core/PyByteArray.java]: This comment seems to belong in #119, where you'll find we already agree about the problem.

The comment about "8-bit clean" refers to PyString objects in which a stored `char` exceeds code point 255, not `itemsize`.

`bytearray` in-place concatenation `+=` takes any buffer object. [Look](https://github.com/python/cpython/blob/0d165262d949440e5aea6533b10e19e4cd5cf12d/Objects/bytearrayobject.c#L339). Here we need only guard _against_ types that can't give you a simple buffer `PyUnicode` (which `setslice` does for you).

**Comment by jeff5** [comment]
The way that #119 was solved is the way to make `memoryview` (and anything else offering bytes via with the buffer interface) acceptable to `bytearray.__iadd__`. Thanks to work on #119, we now have a clean way to get at those bytes using the idiom:
```
try (PyBuffer buf = ((BufferProtocol) obj).getBuffer(PyBUF.SIMPLE)) {
    // ... process those bytes .... 
} catch (ClassCastException e) {
    throw Py.TypeError("expected string or buffer, but got " + obj.getType());
}
```
@devendor : were you interested in finishing this? If not, it was still a useful contribution that you identified the problems in #119 and #120.

**Comment by jeff5** [comment]
I believe this is covered now by #134 but thanks @devendor for raising this and for your contribution to our thinking.

**Pull request closed without merging**